liver_imaging_analysis/engine/engine.py

```diff
@@ -428,62 +536,166 @@ def test(self, dataloader = None, callback = False):
                 batch = self.post_process(batch)
                 self.metrics(batch[Keys.PRED].int(), batch[Keys.LABEL].int())
                 if callback:
-                  self.per_batch_callback(
-                      batch_num,
-                      batch[Keys.IMAGE],
-                      batch[Keys.LABEL],
-                      batch[Keys.PRED]
-                      )
+                    self.per_batch_callback(
+                        batch_num,
+                        batch[Keys.IMAGE],
+                        batch[Keys.LABEL],
+                        batch[Keys.PRED],
+                    )
             test_loss /= num_batches
             # aggregate the final metric result
             test_metric = self.metrics.aggregate().item()
             # reset the status for next computation round
             self.metrics.reset()
         return test_loss, test_metric
 
+    def objective(
+        self,
+        trial,
+        automate,
+        pretrained,
+        cp_path,
+        epochs,
+        evaluate_epochs,
+        batch_callback_epochs,
+        save_weight,
+        save_path,
+        test_batch_callback,
+    ):
+        """
+        Train (by manually selecting hyperparameters or utilizing Optuna).
+
+        Parameters
+        ----------
+        trail: int
+                Describes number of trials of each hyperparameter combination
+        automate: bool
+                Flag to use automatic hyperparameter optimization.
+        pretrained : bool
+            if true, loads pretrained checkpoint. Default is True.
+        cp_path : str
+            determines the path of the checkpoint to be loaded
+            if pretrained is true. If not defined, the potential
+            cp path will be loaded from config.
+        epochs : int
+            number of training epochs.
+            If not defined, epochs will be loaded from config.
+        evaluate_epochs : int
+            The number of epochs to evaluate model after. Default is 1.
+        batch_callback_epochs : int
+            The frequency at which per_batch_callback will be called.
+            Expects a number of epochs. Default is 100.
+        save_weight : bool
+            whether to save weights or not. Default is True.
+        save_path : str
+            the path to save weights at if save_weights is True.
+            If not defined, the potential cp path will be loaded
+            from config.
+        test_batch_callback : bool
+            whether to call per_batch_callback during testing or not.
+            Default is False
+
+        Returns
+        -------
+        float
+            the averaged loss calculated during validation
+
+        """
+
+        if automate == True:
+
+            config.network_parameters["num_res_units"] = trial.suggest_int(
+                "res_units_l{}", 2, 5
+            )
+            config.training["optimizer"] = trial.suggest_categorical(
+                "optimizer", ["Adam", "SGD"]
```

Comment:
Maybe try to make this function more generic by selecting candidates from the list of supported optimizers? For example, if we add a new optimizer later, this function automatically sets it as one of the candidates. Same for other parameters.